[Fix  #4333] Implement asynchronous search reindex functionality using celery

safwanrahman opened the pull request:
The management command `reindex_elasticsearch` has been rewritten from scratch using celery tasks.
The idea is taken from @robhudson [blog post](http://rob.cogit8.org/blog/2014/11/using-celerys-canvas-to-reduce-elasticsearch-index-time/) and heavily inspired from the code of [mozilla/zamboni](https://github.com/mozilla/zamboni/blob/master/lib/es/management/commands/reindex.py).

Need to overwrite some method of `django-elasticsearch-dsl` in order to support zero downtime rebuild (https://github.com/sabricot/django-elasticsearch-dsl/issues/75). I am hoping to send a PR to upstream.

This fixes  #4333

@ericholscher @rtfd/core r?

safwanrahman commented:
I have tested with about 10K files in single celery host with 4 workers. The indexing got finished within 20 seconds.

ericholscher commented on readthedocs/search/management/commands/reindex_elasticsearch.py:
```diff
@@ -0,0 +1,95 @@
+    @staticmethod
+    def _get_models(args):
+        for model_name in args:
+            yield apps.get_model(model_name)
```
This doesn't need to be broken out into it's own function. It just increases complexity for little value.

ericholscher commented on readthedocs/search/management/commands/reindex_elasticsearch.py:
```diff
@@ -0,0 +1,95 @@
+                  "The format is <app_label>.<model_name>")
+        )
+
+    def handle(self, *args, **options):
```
Needs a docstring showing how to run it.

ericholscher commented on readthedocs/search/mixins.py:
```diff
@@ -0,0 +1,66 @@
+            ),
+        }
+
+    def _get_actions(self, object_list, action, index_name=None):
```
Why are we overriding this? Needs a docstring.

ericholscher commented on readthedocs/search/tasks.py:
```diff
@@ -0,0 +1,84 @@
+
+    for document in documents:
+        if str(document) == document_class:
+            return document
```
What is the logic here for? Will there be multiple documents for a model ever? Also needs a docstring :)

safwanrahman replied:
Yes. there can be multiple documents for a model

```python
class FooDocument(DocType):
    ...
    class Meta:
        model = Bar

class NewFooDocument(DocType):
    ...
    class Meta:
        model = Bar
```

ericholscher commented on readthedocs/search/management/commands/reindex_elasticsearch.py:
```diff
@@ -0,0 +1,95 @@
+    def _run_reindex_tasks(self, models):
+        for doc in registry.get_documents(models):
+            qs = doc().get_queryset()
+            instance_ids = list(qs.values_list('id', flat=True))
```
Is this not going to use a ton of memory? This is creating the entire list of objects in memory all at once, instead of streaming them.

safwanrahman replied:
Its not actually creating list of objects. Its just a list of integers. I think integers take much lower amount of memory.

ericholscher replied:
I'm pretty worried this is going to be both slow & perhaps incredibly memory intensive in production when we have 2 million objects in memory here. I'd like to find another approach that streams the data if possible, but we can try this for now and see what happens.

safwanrahman replied:
@ericholscher I think we can try this now to check if anything wrong happens.

safwanrahman replied:
@ericholscher I have optimized the memory usage in 143ce7f

ericholscher commented on readthedocs/search/tasks.py:
```diff
@@ -0,0 +1,84 @@
+
+
+@app.task(queue='web')
+def switch_es_index_task(app_label, model_name, index_name, new_index_name):
```
We don't put `task` at the end of our tasks.

safwanrahman replied:
So what do we put to understand that its a task?

ericholscher replied:
We don't use any `suffix` in our code. We should be keeping the same coding standards here as elsewhere. If we want to change this, we need to change it everywhere, otherwise it's even more confusing having half of our tasks end in `task`.

safwanrahman replied:
Yeah. Fixed it in latest commit

ericholscher commented:
Another thought -- should this be implemented as a contribution to DED with a Celery backend, instead of our own custom logic? It sounds like we might be able to do it fix Celery with a setting: https://github.com/sabricot/django-elasticsearch-dsl#elasticsearch_dsl_signal_processor

safwanrahman commented:
>Another thought -- should this be implemented as a contribution to DED with a Celery backend, instead of our own custom logic? It sounds like we might be able to do it fix Celery with a setting:
https://github.com/sabricot/django-elasticsearch-dsl#elasticsearch_dsl_signal_processor

Sure, I will surely port it to `DED`, but I wonder if we can use the signal processor as that is called while saving/updating the objects.

> We could also perhaps add a celery flag to the existing indexer command? So that we don't have to maintain our own set of code around indexing.

Thats really true. But it needs some time to get it reviewed and merged into master. I will open a PR there soon, but to get our deployment soon, we can keep it here untill then.

ericholscher commented:
> Thats really true. But it needs some time to get it reviewed and merged into master. I will open a PR there soon, but to get our deployment soon, we can keep it here untill then.

We can always deploy a forked version while we wait for it to get accepted. It's true though we don't want to be waiting for review & merge from them, but perhaps it will be quick.

safwanrahman commented:
> I'm a little concerned about the complexity of this approach. Is there a reason we're using chunks here when we already have a domain object to chunk on which is the Version? This feels like extra work to do and code to maintain, when we already have an existing way to think about this problem.

This is a general purpose management command for indexing all type of documents like `Project`/`HTMLFile`, not only the `HTMLFile`. So I had only one way to make a chunk depending on number.

> This approach also doesn't use the same code path for indexing as our application code, so now we have two different ways of indexing files, which doesn't seem great.

Yeah, thats true. This management command actually reindex all the documents. On the otherside, `DED` catch the signal when a new file is created and index it into Elasticsearch.

> It really feels like all we need is:

    - A management commands that iterates over versions, and sends them to be indexed
    - A celery task that takes a version and indexes all the files in that version, which is called both in production as well as in reindexing.

As mentioned above, the management command is general purpose. So something specially for `HTMLFile` would be extra implementation that maybe not available when we port it to `DED`

ericholscher commented:
> Yeah, thats true. This management command actually reindex all the documents. On the otherside, DED catch the signal when a new file is created and index it into Elasticsearch.

Right, but that doesn’t scale properly right now. It currently sends an http request per file saved during project build. We need to batch it, which could use similar logic to this.

safwanrahman commented:
> Right, but that doesn’t scale properly right now. It currently sends an http request per file saved during project build. We need to batch it, which could use similar logic to this.

I understand. but how do we get the files in batch? Do there any signal that is sent in batch?

ericholscher commented:
We should just update the code to keep track of the files that are changed, as it works now.

safwanrahman commented:
> Do we need this to be another Celery task? It's already running inside of Celery, so I'm not sure we need to delay it here. It could lead to weird race conditions if it got queued up for a long time.

Good catch @ericholscher. I will fix it in synchronous task. Thanks

safwanrahman commented:
@ericholscher I have fixed the issues as you mentioned and fixed the tests. Also added a comment in https://github.com/rtfd/readthedocs.org/issues/4264#issuecomment-406404957 for testing it in backlog. r?

ericholscher commented on readthedocs/search/management/commands/reindex_elasticsearch.py:
```diff
@@ -7,69 +7,73 @@
+            queryset = doc().get_queryset()
+            # Get latest object from the queryset
+            latest_object = queryset.latest('modified_date')
+            latest_object_time = latest_object.modified_date
```
This should just use the current time, not query the entire queryset for the time. This will be quite a slow query likely in prod.

safwanrahman replied:
Yeah! I was also thinking about it.

safwanrahman commented:
@ericholscher Fixed the `timestamp` issue. r?

ericholscher commented:
This looks good. 👍
